### PR TITLE
add type annotations for config

### PR DESCRIPTION
**Pull request by joshunrau**
I added a Config class which defines the structure of the config object to be passed as argument to extract_UKBB_tabular_data. This does not change any runtime behavior, but makes it easier for developers to understand the codebase (and allows for static analysis with tools such as MyPy.

### CODE REVIEW

**Comment by gdevenyi** [comment]
Can you please run the code through `black`

File: melted_UKBB_extract.py [hunk 1]
```diff
@@ -329,16 +335,7 @@ def extract_UKBB_tabular_data(
...
-        with open(args.config_file, "r") as stream:
-            try:
-                config = yaml.safe_load(stream)
-            except yaml.YAMLError as exc:
-                logging.exception(exc)
-                sys.exit(1)
-    except FileNotFoundError as exc:
-        logging.exception(exc)
-        sys.exit(1)
+    config = Config.from_yaml(args.config_file)
```
Comment (gdevenyi) [inline melted_UKBB_extract.py:338]: Switching to this implementation means you can remove the `import yaml` right?

File: config.py [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    Categories: list[Any]
+    recode_field_names: bool
+    recode_data_values: bool
+    drop_empty_strings: bool
+    drop_extra_NA_codes: bool
+    wide: bool
+    recode_wide_column_valuetypes: bool
+    convert_compound_to_list: bool
+    convert_less_than_value_integer: Any
+    convert_less_than_value_continuous: Any
```
Comment (gdevenyi) [inline config.py:28]: Can you replicate the comments from the yaml template here?

**Comment by gdevenyi** [comment]
Changes are required to README if this change is made.

File: config.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    ArrayIDs: list[int] | list[None]
+    Categories: list[Any]
+    recode_field_names: bool
+    recode_data_values: bool
+    drop_empty_strings: bool
+    drop_extra_NA_codes: bool
+    wide: bool
+    recode_wide_column_valuetypes: bool
+    convert_compound_to_list: bool
+    convert_less_than_value_integer: Any
```
Comment (gdevenyi) [inline config.py:27]: This must be an integer.

File: config.py (hunk 2 above)
Comment (gdevenyi) [inline config.py:28]: This must be a numeric

File: melted_UKBB_extract.py (hunk 1 above)
Comment (joshunrau) [reply melted_UKBB_extract.py:338]: Yes, that is correct

File: config.py (hunk 2 above)
Comment (joshunrau) [reply config.py:28]: I typed it as `int | float`

File: melted_UKBB_extract.py (hunk 1 above)
Comment (joshunrau) [reply melted_UKBB_extract.py:341]: c71b2fe2519f596f0abf25b3bfe990e03f0bb075

File: config.py (hunk 2 above)
Comment (joshunrau) [reply config.py]: a633e80e139fe47c408944ad4ca7933d568f2ddf

File: config.py (hunk 3 above)
Comment (joshunrau) [reply config.py]: 0da710d4cfd04ddcd5c058220abdd2b13ff40122

**Comment by gdevenyi** [comment]
> > Changes are required to README if this change is made.
> 
> What kind of changes?

The README describes the pythonic interface fairly generally, these changes improve that usage, worth describing how to use that.

File: requirements.txt [hunk 4]
```diff
@@ -0,0 +1,2 @@
+polars==0.17.2
+PyYAML==6.0
```
Comment (gdevenyi) [inline requirements.txt:2]: Please change to >=

**Review by gdevenyi — CHANGES_REQUESTED** [review]
Can you please update this part of the README https://github.com/CoBrALab/UKBB-tabular-processing#use-inside-python given the changes to the interface.

File: README.md [hunk 5]
```diff
@@ -56,18 +56,36 @@ $ python melted_UKBB_extract.py --config-file myconfig.yaml --data-file current.
...
+    config: Config,
+    data_file: str | None = None,
+    dictionary_file: str | None = None,
+    coding_file: str | None = None,
+    verbose: str | None = False,
+) -> tuple[pl.DataFrame, pl.DataFrame | None, pl.DataFrame, pl.DataFrame]:
 ```
 
-The function returns 3 or 4 polars DataFrames depending on the `config['wide']` setting:
+The return signature depends on the `config['wide']` setting. Unfortunately, as Python < 3.11 does not support `TypedDict` with generic types, this cannot be (easily) expressed with Python hints. Once Python 3.10 is no longer supported, this could be updated with a generic type.
```
Comment (gdevenyi) [inline README.md:71]: Just to be clear here, TypedDict is supported, but not the full set of features we would like to use <3.11?

File: melted_UKBB_extract.py [hunk 6]
```diff
@@ -3,20 +3,22 @@
...
-    import polars as pl
-    import pathlib as p
+    data_file: str,
+    dictionary_file: str,
+    coding_file: str,
+    verbose: Optional[bool] = False,
+) -> tuple[pl.DataFrame, pl.DataFrame | None, pl.DataFrame, pl.DataFrame]:
 
-    pl.Config.set_verbose(verbose)
+    pl.Config.set_verbose(verbose == True)
```
Comment (gdevenyi) [inline melted_UKBB_extract.py:21]: why this change? can verbose not be passed directly into this?

**Review by gdevenyi — CHANGES_REQUESTED** [review]

File: melted_UKBB_extract.py (hunk 6 above)
Comment (joshunrau) [reply melted_UKBB_extract.py:21]: According to MyPy, this method only accepts `bool`. It is possible `None` would work at runtime, but why not be safe and make sure it is a `bool`.

Comment (gdevenyi) [reply melted_UKBB_extract.py:21]: why would it ever be none? the function call had  a default value of  `verbose = False`

Comment (joshunrau) [reply melted_UKBB_extract.py]: yup, this is gone now.

Comment (joshunrau) [reply melted_UKBB_extract.py]: 1a16e48edd1db5c39e8a6b909a7502861abb8cfe
